documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,366 @@
+  - Atan2Pi(Vector x, Vector y)
+  - Atan2(Vector x, TScalar y)
+  - Atan2Pi(Vector x, TScalar y)
+  - BitDecrement(Vector x)
+  - BitIncrement(Vector x)
+  - FusedMultiplyAdd(Vector left, Vector right, Vector addend)
+  - FusedMultiplyAdd(Vector left, Vector right, TScalar addend)
+  - FusedMultiplyAdd(Vector left, TScalar right, Vector addend)
+  - FusedMultiplyAdd(Vector left, TScalar right, TScalar addend)
+  - ReciprocalEstimate(Vector x)
+  - ReciprocalSqrtEstimate(TVector x)
+  - ILogB(Vector x)
+    - Returns VectorNI<T>, where N matches the dimensionality of the vector
+    - **INFORMATIVE** This may require multiple methods depending on implementation
+  - ScaleB(Vector x, VectorNI<int> n)
+  - ScaleB(Vector x, int n)
+  - RoundToInt(Vector x)
+    - Returns `VectorNI`, where N matches the dimensionality of the vector
+    - **INFORMATIVE** This may require multiple methods depending on implementation
+  - FloorToInt(Vector x)
+    - Returns `VectorNI`, where N matches the dimensionality of the vector
+    - **INFORMATIVE** This may require multiple methods depending on implementation
+  - CeilingToInt(Vector x)
+    - Returns `VectorNI`, where N matches the dimensionality of the vector
+    - **INFORMATIVE** This may require multiple methods depending on implementation
+  - ToVector64(Vector x)
+    - Returns `System.Runtime.Intrinsics.Vector64<TScalar>`
+  - ToVector128(Vector x)
+    - Returns `System.Runtime.Intrinsics.Vector128<TScalar>`
+  - ToVector256(Vector x)
+    - Returns `System.Runtime.Intrinsics.Vector256<TScalar>`
+  - ToVector512(Vector x)
+    - Returns `System.Runtime.Intrinsics.Vector512<TScalar>`
+
+# Matrix Types
+
+This proposal includes the following matrix types:
+- Matrix2x2F
+- Matrix2x2I
+- Matrix2x3F
+- Matrix2x3I
+- Matrix2x4F
+- Matrix2x4I
+- Matrix3x2F
+- Matrix3x2I
+- Matrix3x3F
+- Matrix3x3I
+- Matrix3x4F
+- Matrix3x4I
+- Matrix4x2F
+- Matrix4x2I
+- Matrix4x3F
+- Matrix4x3I
+- Matrix4x4F
+- Matrix4x4I
+- Matrix5x4F
+- Matrix5x4I
+
+Integer Variants do not require any functions which interact with Quaternions
+
+Matrix structs **must** fulfill the following requirements:
+- Fulfills `IEquatable<T>` where `T` is the same matrix class
+- Stored in row major format
+- F matricies work with F vectors, and I Matricies work with I vectors
+- Both row vectors and individual values (M11, etc.) accessible via properties
+- A ref indexer that takes row and column indicies and outputs the value
+- Add, subtract, and multiply operators defined with Matricies of the same size
+- Multiply operators defined with compatible matricies, if the output matrix type already exists (AxB * BxC = AxC)
+- Multiply operator defined with the appropriate Vector type (vector4F * Matrix4x4F is valid)
+  - **optionally**, include multiplication operators for Vector types smaller than the appropriate type, assuming 1 for the missing dimensions (vector2F * Matrix4x4F -> (Vector2F.x, Vector2F.y, 1, 1) * Matrix4x4F)
```

Comment:
this is specified required in the vector types above. additionally, im unsure if stepping up multiple dimensions should be permissible

Reply:
removed from here, but I did agree that this functionality should match Transform as Petre mentioned in the discord. I could go either way, but it was the decision I made at the time.